### PR TITLE
Adding ReciprocalRank transformation

### PR DESCRIPTION
**Pull request by lastmansleeping**
In this PR, we add support for a keras layer and feature function that can be used to convert raw scores into reciprocal rank scores.

### CODE REVIEW

File: python/ml4ir/base/features/preprocessing.py [hunk 1]
```diff
@@ -130,6 +133,8 @@ def preprocess_text(
     Output:
         >>> "abcabc123"
     """
+    if remove_quotes:
```
Comment (lastmansleeping) [inline python/ml4ir/base/features/preprocessing.py:136]: This is another minor change I strangely had in my git stash. Can move to another PR if needed.

**Review by marjanhs — COMMENTED** [review]
I left some comments mostly about the structure and questions.

File: python/ml4ir/applications/ranking/tests/test_normalization.py [hunk 2]
```diff
@@ -63,11 +64,62 @@ def test_tmm_norm_with_2d_feature(self):
...
 
         self.assertTrue(np.isclose(actual_normed_feature, expected_normed_feature).all())
         self.assertTrue(input_feature.shape == actual_normed_feature.shape)
+
+
+class TestReciprocalRank(unittest.TestCase):
+
+    def test_reciprocal_rank_default(self):
+        reciprocal_rank_op = ReciprocalRank()
+        input_feature = np.array([[0.0, 0.8, 0.9, 0.5, 0.6, 0.7]])
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/tests/test_normalization.py:84]: Good point. I guess the scores could be negative too, you are right. Would need to use `ignore_zero_score = False` in that case though. Will add a note.

File: python/ml4ir/applications/ranking/features/feature_fns/normalization.py [hunk 3]
```diff
@@ -59,4 +60,66 @@ def call(self, inputs, training=None):
...
+                Use zero reciprocal rank for score value of 0.0
+        """
+        super().__init__(feature_info=feature_info, file_io=file_io, **kwargs)
+
+        self.reciprocal_rank_op = ReciprocalRankLayer(
+            k=self.feature_layer_args.get(self.K, 0.),
+            k_trainable=self.feature_layer_args.get(self.K_TRAINABLE, False),
+            ignore_zero_score=self.feature_layer_args.get(self.IGNORE_ZERO_SCORE, True))
+
+    def call(self, inputs, training=None):
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/features/feature_fns/normalization.py:107]: Yes, `None` gets translated to `False` I believe. This is just the standard tensorflow keras layer function signature though.

File: python/ml4ir/applications/ranking/features/feature_fns/normalization.py [hunk 4]
```diff
@@ -1,7 +1,9 @@
 import tensorflow as tf
 
 from ml4ir.base.features.feature_fns.base import BaseFeatureLayerOp
-from ml4ir.applications.ranking.model.layers.normalization import TheoreticalMinMaxNormalization as TheoreticalMinMaxNormalizationLayer
+from ml4ir.applications.ranking.model.layers.normalization import \
+    TheoreticalMinMaxNormalization as TheoreticalMinMaxNormalizationLayer
+from ml4ir.applications.ranking.model.layers.normalization import ReciprocalRank as ReciprocalRankLayer
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/features/feature_fns/normalization.py:6]: Sure, yes. I can do that. One is a keras layer that is generic and can be used with any tensorflow, the other is a feature transformation function (which is just a wrapper in this case) that is ml4ir specific.

File: python/ml4ir/applications/ranking/model/layers/normalization.py [hunk 5]
```diff
@@ -122,4 +122,83 @@ def call(self, inputs, training=None):
             tf.math.subtract(inputs, self.theoretical_min),
             tf.math.subtract(query_max, self.theoretical_min))
 
-        return normed_inputs
\ No newline at end of file
+        return normed_inputs
+
+
+class ReciprocalRank(layers.Layer):
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/model/layers/normalization.py:128]: I considered that. Converting scores to ranks seemed similar enough to normalization in some ways to me. But I can move it to a separate module, yes.

File: python/ml4ir/base/features/feature_layer.py [hunk 6]
```diff
@@ -35,6 +36,7 @@ def __init__(self):
             # Ranking based feature transforms
             CategoricalVector.LAYER_NAME: CategoricalVector,
             TheoreticalMinMaxNormalization.LAYER_NAME: TheoreticalMinMaxNormalization,
+            ReciprocalRank.LAYER_NAME: ReciprocalRank,
             QueryLength.LAYER_NAME: QueryLength,
             QueryTypeVector.LAYER_NAME: QueryTypeVector
```
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/base/features/feature_layer.py:41]: This specific feature transform, yes. We can use it for other apps too, just that semantically they made more sense to be in ranking.

File: python/ml4ir/applications/ranking/model/layers/normalization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lastmansleeping) [reply python/ml4ir/applications/ranking/model/layers/normalization.py:128]: Do you have any suggestions for the name?

File: python/ml4ir/applications/ranking/tests/test_rank_transform.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    def test_reciprocal_rank_ignore_zero_score(self):
+        reciprocal_rank_op = ReciprocalRankLayer(ignore_zero_score=False)
+        input_feature = np.array([[0.0, 0.8, 0.9, 0.5, 0.6, 0.7]])
+
+        actual_reciprocal_ranks = reciprocal_rank_op(input_feature).numpy()
+
+        expected_reciprocal_ranks = np.array([[1. / 6, 1. / 2, 1. / 1, 1. / 5, 1. / 4, 1. / 3]])
+        self.assertTrue(np.isclose(actual_reciprocal_ranks, expected_reciprocal_ranks).all())
+
+    def test_reciprocal_rank_negative_scores(self):
```
Comment (lastmansleeping) [inline python/ml4ir/applications/ranking/tests/test_rank_transform.py:55]: @marjanhs Added test for negative scores

**Review by marjanhs — APPROVED** [review]
LGTM.
